server/ft.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2017 Apcera Inc. All rights reserved.
+
+package server
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/nats-io/go-nats"
+	"github.com/nats-io/nats-streaming-server/spb"
+	"github.com/nats-io/nats-streaming-server/stores"
+	"github.com/nats-io/nats-streaming-server/util"
+)
+
+// FT constants
+const (
+	ftDefaultHBInterval       = time.Second
+	ftDefaultHBMissedInterval = 1250 * time.Millisecond
+)
+
+var (
+	// Some go-routine will panic, which we can't recover in test.
+	// So the tests will set this to true to be able to test the
+	// correct behavior.
+	ftNoPanic bool
+	// For tests purposes, we may want to pause for the first
+	// attempt at getting the store lock so that test can
+	// switch store with a mocked one.
+	ftPauseBeforeFirstAttempt bool
+	ftPauseCh                 = make(chan struct{})
+	// This can be changed for tests purposes.
+	ftHBInterval       = ftDefaultHBInterval
+	ftHBMissedInterval = ftDefaultHBMissedInterval
+)
+
+func ftReleasePause() {
+	ftPauseCh <- struct{}{}
+}
+
+// ftStart will return only when this server has become active
+// and was able to get the store's exclusive lock.
+// This is running in a separate go-routine so if server state
+// changes, take care of using the server's lock.
+func (s *StanServer) ftStart() (retErr error) {
+	Noticef("STAN: Starting in standby mode")
+	// For tests purposes
+	if ftPauseBeforeFirstAttempt {
+		<-ftPauseCh
+	}
+	print, _ := util.NewBackoffTimeCheck(time.Second, 2, time.Minute)
+	for {
+		select {
+		case <-s.ftQuit:
+			// we are done
+			return nil
+		case <-s.ftHBCh:
+			// go back to the beginning of the for loop
+			continue
+		case <-time.After(ftHBMissedInterval):
+			// try to lock the store
+		}
+		locked, err := s.ftGetStoreLock()
+		if err != nil {
+			// TODO: This means that we got an error not related to locking.
+			// Since we are in standby, should just keep trying? It could
```

Comment:
Remember that `store.GetExclusiveLock()` is a store API, so depending on the store implementation, Permission or Path doesn't exist may not make sense. We should then specify at the Store.GetExclusiveLock() the behavior, for instance return error if failing the get the lock is considered fatal as opposed to maybe temporary? If that's the case, the changes would be in this PR in filestore and here I would indeed return error if `err != nil`.